### PR TITLE
Refactoring and adding some negative test cases

### PR DESCRIPTION
**Pull request by lovesh**
Adding negative tests where the tested agent returns an error message when it can.
The PR might look like huge but a lot of it is fixing the imports. I have changed the package name so that its a valid package name and made some directories packages that can be used in imports.

### CODE REVIEW

**Comment by lovesh** [comment]
@TelegramSam I don't follow. Can you please point to the code in the PR that does the change?

**Comment by dhh1128** [comment]
@lovesh, wouldn't this line be an example of what Sam is talking about? https://github.com/hyperledger/indy-agent/blob/d7bd8a4459571ff5a769f4c68e5e807b99104783/python/agent.py#L13

We are now having to import from "python.something" instead of just importing "something" -- so the overall python package has now been moved up a level.

**Comment by dhh1128** [comment]
The most important thing about this PR is the introduction of some new tests; the cleanup is a separate issue.

**Comment by lovesh** [comment]
@dhh1128 I made `python` folder a package so that i can run it from outside. In my experience having a package avoid import issues. The other option would be to move scripts like agent.py and indy-agent.py in a scripts folder.

**Comment by TelegramSam** [comment]
Untangling the ref agent from the testing code looks good, and resolves my issue.

>I made python folder a package so that i can run it from outside. In my experience having a package avoid import issues. 
I have no problem with making the python folder a package. I _do_ have an issue with changing all the imports to use that package name, as it forces everybody to run the agent with a base path of indy-agent, instead of indy-agent/python.

We have multiple parties using the code with the current package organization. Making this change at the same time as the others in the commit makes it difficult to accept. If you feel the changes must be made at the same time, please further explain your reasoning. If they do not need to be made at the same time, please move them to another PR so we can have that conversation without slowing down the acceptance of the new test cases.

**Comment by lovesh** [comment]
@TelegramSam Changed the imports

**Comment by TelegramSam** [comment]
Running tests failed. Still digging through this, but if you have any insight that can help me, I can get this merged faster:
```(sovrin) root@indy-agent-dev-2:/media/indy-agent/test_suite# pytest
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-3.6.4, py-1.7.0, pluggy-0.7.1 -- /opt/sovrin/bin/python3
cachedir: .pytest_cache
rootdir: /media/indy-agent/test_suite, inifile: pytest.ini
plugins: asyncio-0.9.0
collected 0 items / 3 errors                                                   

==================================== ERRORS ====================================
______________________ ERROR collecting connection.manual ______________________
ImportError while importing test module '/media/indy-agent/test_suite/tests/connection/manual.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/__init__.py:12: in <module>
    from message import Message
E   ModuleNotFoundError: No module named 'message'
_____________________ ERROR collecting basicmessage.manual _____________________
ImportError while importing test module '/media/indy-agent/test_suite/tests/basicmessage/manual.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/__init__.py:12: in <module>
    from message import Message
E   ModuleNotFoundError: No module named 'message'
______________________ ERROR collecting trustping.manual _______________________
ImportError while importing test module '/media/indy-agent/test_suite/tests/trustping/manual.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/__init__.py:12: in <module>
    from message import Message
E   ModuleNotFoundError: No module named 'message'
!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 3 error in 0.14 seconds ============================```

**Comment by lovesh** [comment]
Since the python package is not in python path, pytest is unable to find it. I was running the tests on Pycharm and i had the `python` directory marked as source.

**Comment by lovesh** [comment]
Adding the `python` package to Python path and then running pytest works. This is what i do.
```
## In test_suite directory
export PYTHONPATH=/home/lovesh/dev/evernym/indy-agent/python:$PYTHONPATH
pytest
```

**Comment by TelegramSam** [comment]
why is that necessary given your code changes? It works without the path change prior to your commits. Why is a python path change for a test suite a good idea?

**Comment by TelegramSam** [comment]
To further clarify: At present, we want absolutely no dependencies between the test suite and the python ref agent. Being redundant in this case is the desired path.

**Comment by lovesh** [comment]
> we want absolutely no dependencies between the test suite and the python ref agent

Can you please explain why?

The test suite is supposed to check ideal functionality and the ref agent is supposed to implement that ideal functionality. eg. The validation rules used in ref agent should be the ones checked in test-suite.

**Comment by dhh1128** [comment]
@lovesh The test suite isn't a test of the ref agent. It's a test of *any* agent. The ref agent isn't guaranteed to be the one that has the most compatibility to the test suite, and it's perfectly reasonable for it not to pass some of the tests.

We are not certain that the ref agent is going to continue to exist in its current form. It may be superseded by the python agent in the indy-cat codebase from BC Gov. Thus, the test suite should not be assuming anything about the ref agent.

**Comment by lovesh** [comment]
@dhh1128 I understood that the test suite is for **any agent** but was assuming the ref agent will always pass the test suite as it was the *reference*. Will remove the assumption from the codebase

File: python_agent_utils/messages/connection.py [hunk 1]
```diff
@@ -96,31 +100,30 @@ def build(label: str, my_did: str, my_vk: str, endpoint: str) -> Message:
...
-                    'did_doc'
+                    DIDDoc.DID,
+                    DIDDoc.DID_DOC
                 ],
-                request['connection']
+                request[Connection.CONNECTION]
             )
 
-            Connection.DIDDoc.validate(request['connection']['did_doc'])
+            DIDDoc.validate(request[Connection.CONNECTION][DIDDoc.DID_DOC])
```
Comment (dbluhm) [inline python_agent_utils/messages/connection.py]: `request[Connection.CONNECTION][DIDDoc.DID_DOC]` is not a `Message` so the `did_doc.validate` in the first line of `DIDDoc.validate` is failing with "dict has no attribute 'validate'"

File: python/modules/__init__.py [hunk 2]
```diff
@@ -1,2 +1,13 @@
...
+from python_agent_utils.messages.message import Message
+
+
+class Module:
+    PROBLEM_REPORT = 'problem_report'
+
+    @staticmethod
+    def build_problem_report_for_connections(family, problem_code, problem_str):
+        return Message({
+            "@type": "{}/problem_report".format(family),
```
[the remark this answers is not in the document]
Comment (lovesh) [reply python/modules/__init__.py]: Changed. Thanks

File: python_agent_utils/messages/connection.py (hunk 1 above)
Comment (lovesh) [reply python_agent_utils/messages/connection.py]: Changed. Thanks

**Comment by lovesh** [comment]
@dbluhm 
> To run this as I normally would from the command line, I had to alter the PYTHONPATH as you mentioned in a previous comment.

From which directory are you running tests from? I run tests from `indy-agent/test_suite` and don't have to modify PYTHONPATH

**Comment by dbluhm** [comment]
> From which directory are you running tests from? I run tests from `indy-agent/test_suite` and don't have to modify PYTHONPATH

Not sure if I was in a bad state before or something but retrying from the test_suite directory, I did not have to modify `PYTHONPATH`. To get the reference agent to run, however, I still did have to modify the `PYTHONPATH` when running from within the python directory.

From within the python directory:
```
$ PYTHONPATH=../:$PYTHONPATH python indy-agent.py --wallet alice test --ephemeralwallet 3001
```

**Comment by lovesh** [comment]
@dbluhm Yes. You are correct. I was running the agent with my IDE which adds source directories to python path. But when in run it from terminal i do have to add indy-agent to PYTHONPATH since `python_agent_utils` is outside `python`. We can either make `python_agent_utils` installable as pip or add 2 lines at top of indy-agent.py
```
import sys
sys.path.append('..')
```

I find the second option better for now.

**Comment by lovesh** [comment]
@dbluhm Thanks for accepting it.

**Pull request merged**